RUBY 2713 wex charging bo 3 1 5 reverse a payment page

jjromeo opened the pull request:
- [RUBY-2713] Add functionality for reversing payments in waste exemptions engine
- [RUBY-2713] Update waste_exemptions_engine and enhance record reversal functionality


[RUBY-2713]: https://eaflood.atlassian.net/browse/RUBY-2713

jjromeo committed:
[RUBY-3414] Add payment details page and link to registrations

- Created `PaymentDetailsController` with an `index` action to display payment details.
- Added a new view `index.html.erb` for the payment details page.
- Updated `RegistrationsController` to remove the `authorize` method.
- Enhanced `ActionLinksHelper` with `display_payment_details_link_for?` method.
- Modified `_resource_details.html.erb` to include current balance and link to payment details.
- Updated locale files

jjromeo committed:
[RUBY-3414] Add conditional display for balance in resource details view

Wrapped the balance display section in a conditional to only show if the `resource` is a `WasteExemptionsEngine::Registration`.

jjromeo committed:
[RUBY-3414] Add safe navigation operator to account balance in resource details view

This change ensures that the `balance` is accessed safely using the safe navigation operator to prevent potential nil errors when `account` is nil.

jjromeo committed:
Merge branch 'main' into RUBY-3414-wex-charging-bo-1-2-view-details-page

jjromeo committed:
[RUBY-3414] Add request specs for payment details view

- Created `payment_details_spec.rb` to test the payment details page.
- Includes tests for rendering the show template with correct registration reference when a user is signed in.
- Includes tests for redirecting to the sign-in page when no valid user is signed in.

jjromeo committed:
Merge branch 'RUBY-3414-wex-charging-bo-1-2-view-details-page' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-3414-wex-charging-bo-1-2-view-details-page

jjromeo committed:
[RUBY-3414] Remove `companies` table from schema.rb

jjromeo committed:
Merge branch 'main' into RUBY-3414-wex-charging-bo-1-2-view-details-page

jjromeo committed:
Merge branch 'main' into RUBY-3414-wex-charging-bo-1-2-view-details-page

jjromeo committed:
[RUBY-3414] Remove unused `view_certificate_token` and `view_certificate_token_created_at` columns from schema.

jjromeo committed:
Merge branch 'RUBY-3414-wex-charging-bo-1-2-view-details-page' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-3414-wex-charging-bo-1-2-view-details-page

jjromeo committed:
[RUBY-2704] Add refund payment functionality

- Implement `RecordRefundFormsController` to handle refund operations, including `index`, `new`, and `create` actions.
- Create `RecordRefundForm` model with validation and submission logic for recording refunds.
- Introduce `RecordRefundService` to process refunds and update account balances.
- Add views for listing payments eligible for refunds and for creating a refund.
- Update `FinanceDetailsHelper` to adjust currency display logic.
- Modify routes to include refund paths under

jjromeo committed:
Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page

jjromeo committed:
[RUBY-2704] Add 'Refund a Payment' feature to the Waste Exemptions back-office

- Updated `Gemfile` and `Gemfile.lock` to point to the `RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page` branch of the `waste_exemptions_engine` for the new refund feature.
- Modified `payment_details_controller.rb` to use a new `PaymentDetailsPresenter` for displaying payment details.
- Enhanced `record_refund_form.rb` to include validation for refund amounts, ensuring they do not exceed the payment amount or account balance.
- Created `PaymentDetailsPresenter` to format and present payment, order, and refund details in the views.
- Refactored `payment_details/index.html.erb` to use the presenter for displaying payment details, including orders, payments, refunds, and balance.
- Updated `record_refund_forms/new.html.erb` to use the GOV.UK Design System FormBuilder and display error summaries.
- Added new locale entries in `record_refund_forms.en.yml` and `payment_details.en.yml` for improved internationalization support, including error messages for refund validation.
- Adjusted the database schema to include a `comments` field in the `payments` table and re-ordered some indexes for efficiency.
- Updated `record_refund_forms_spec.rb` to reflect changes in the refund form logic and remove outdated comments.
- This commit introduces a comprehensive refund processing feature, enhancing the back-office capabilities for managing payments and refunds efficiently.

jjromeo committed:
[RUBY-2704] Rubocop fixes

- Refactor `RecordRefundFormsController` to improve code readability and ensure correct method chaining.
- Update `RecordRefundForm` to include validations and refactor methods for better clarity.
- Add localization for success and error messages in `record_refund_forms.en.yml`.
- Use `delegate` in `PaymentDetailsPresenter` for cleaner code and update method chaining.
- Adjust formatting in `FinanceDetailsHelper`.
- Modify request specs in `record_refund_forms_spec

jjromeo committed:
Add `i18n_page` variable to payment details spec

Added the `i18n_page` variable to the `GET /registrations/:reference/payment_details` test context to improve clarity and maintainability of internationalization references.

jjromeo committed:
[RUBY-2704] Refactor `RecordRefundForms` to `RecordRefunds` for consistency

Renamed the `RecordRefundFormsController` to `RecordRefundsController` and updated all related references throughout the application. This includes changes in the controller, views, routes, localization files, and request specs. Adjusted the I18n keys and paths to reflect the new naming convention, ensuring consistency and clarity across the codebase.

jjromeo committed:
Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page

jjromeo committed:
[RUBY-2704] Refactor `reason` to `comments` in refund functionality

Changed the attribute `reason` to `comments` across various files for consistency in refund processing. Updated models, controllers, services, views, and locale files to reflect this change. Adjusted validation and logging accordingly.

jjromeo committed:
[RUBY-2704] Rename `record_refund_forms_controller.rb` to `record_refunds_controller.rb` for consistency

jjromeo committed:
[RUBY-2704] Refactor refund form and update localization keys

- Updated error message keys in `RecordRefundForm` to use a more consistent path.
- Changed the path for new refund forms in views and specs from `new_registration_record_refund_form_path` to `new_registration_record_refund_path`.
- Updated localization files to rename `reason` to `comments` and adjusted error message paths.
- Removed unused `account_id` attribute from the payment factory.
- Updated specs to reflect changes in path and parameter

jjromeo committed:
[RUBY-2704] Update 'comments' to 'reason' in payment details localization and specs

- Changed the label 'comments' to 'reason' in `payment_details.en.yml`.
- Updated corresponding test expectations in `payment_details_spec.rb` to reflect this change.

jjromeo committed:
[RUBY-2704] Remove bang method from `update_balance` call in `RecordRefundService` to prevent raising exceptions.

jjromeo committed:
[RUBY-2704] Remove schema changes as they have been handled already

jjromeo committed:
[RUBY-2704] Update `Gemfile.lock` to latest `waste-exemptions-engine` revision

Updated the `waste-exemptions-engine` gem to revision `aa541fd649a174b96ebfc1466ed7d710fc44e9f0` in the `Gemfile.lock` to incorporate the latest changes on the `RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page` branch.

jjromeo committed:
[RUBY-2704] Add refund functionality enhancements

- Updated `Gemfile.lock` to reflect the latest revision of the waste-exemptions-engine.
- Modified `RecordRefundForm` to include validation for comments presence, ensuring a reason is provided for refunds.
- Added `can_display_refund_link?` method to `PaymentDetailsPresenter` to conditionally display refund links based on account status.
- Enhanced `RecordRefundService` to append "/REFUND" to the payment reference for clarity.
- Updated view logic in

jjromeo committed:
[RUBY-2704] Enhance refund feature and update payment display

- Added `reason_present_in_comments` validation to `RecordRefundForm`.
- Changed date formatting in `PaymentDetailsPresenter` to use `:day_month_year_slashes`.
- Updated payment and refund tables in views to use `created_at` instead of `date_time`.
- Replaced `order_id` with `reference` in the refund index view.
- Displayed payment comments instead of payment type in the refund table.

jjromeo committed:
[RUBY-2704] Add tests for `RecordRefundService` to handle refund processing

This commit introduces a new test suite for the `RecordRefundService` to verify its functionality when processing refunds. It includes tests for successful refund creation, correct attribute assignment, account balance updates, and error handling with logging and Airbrake notification. The tests ensure that the service behaves as expected in both successful and error scenarios.

jjromeo committed:
[RUBY-2704] Update `@payments` to use `refundable` scope in `RecordRefundsController#index` method.

jjromeo committed:
[RUBY-2713] Add functionality for reversing payments in waste exemptions engine

- Updated the `Gemfile` and `Gemfile.lock` to use the local path for the `waste_exemptions_engine` gem instead of the remote git repository, facilitating local development.
- Introduced a new `RecordReversalsController` with actions to handle the index, new, and create operations for reversing payments.
- Added the `RecordReversalForm` class to manage the form data for reversing payments, including validations to ensure the presence of a reason in the comments.
- Implemented the `ReversePaymentService` to encapsulate the logic for processing payment reversals, including creating reversal records, updating the original payment, and adjusting account balances.
- Modified `PaymentDetailsPresenter` to support displaying both refunds and reversals, and added a method to check if a reversal link can be displayed.
- Updated views to include interfaces for listing payments available for reversal and submitting reversal requests.
- Added new locale files and entries (`record_reversals.en.yml`) to support internationalization for the reversal feature.
- Updated routes to include the new reversal routes under the `registrations` resource.
- Updated the database schema to include new fields in the `payments` table for tracking reversals (`reversed_by`, `reversed_at`, `reversal_id`).
- Created request specs (`record_reversals_spec.rb`) and service specs (`reverse_payment_service_spec.rb`) to ensure the functionality of the new reversal feature and its integration with the application.
- Ensured comprehensive error handling and logging within the reversal service, with notifications sent to Airbrake in case of failures.

This commit introduces a complete set of features for reversing payments, enhancing the payment management capabilities of the waste exemptions engine.

jjromeo committed:
[RUBY-2713] Update waste_exemptions_engine and enhance record reversal functionality

- Changed `Gemfile` to use the waste_exemptions_engine from a GitHub repository instead of a local path, specifying the branch `RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page`.
- Updated `Gemfile.lock` to reflect the new source and version of `defra_ruby_validators`.
- Modified `RecordReversalsController` to initialize `Record

jjromeo committed:
[RUBY-2713] Add `user` attribute to `RecordReversalForm`

- Introduced `user` as an attribute in `RecordReversalForm` to remove the need for initialization with a `user` parameter.
- Simplified the initialization process by eliminating the custom `initialize` method.

jjromeo committed:
[RUBY-2704] Refactor `validate_amount` logic in `RecordRefundForm`

Simplified the condition for adding an error when the refund amount exceeds the balance by removing unnecessary checks and directly comparing the amount to the balance.

jjromeo committed:
[RUBY-2704] Enhance refund form validation and add tests

- Added `payment_exists` validation to ensure the payment ID is valid and exists.
- Improved the `submit` method to handle cases where the payment is not found.
- Updated locale file to include error message for missing payment.
- Added RSpec tests for `RecordRefundForm` to validate various scenarios, including invalid amounts, nonexistent payments, and missing comments.

jjromeo committed:
Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page

jjromeo committed:
[RUBY-2704] Refactor refund process and update payment details view

- Updated `Gemfile.lock` with the latest `waste_exemptions_engine` revision.
- Removed account balance update logic from `RecordRefundService` and corresponding test case.
- Refactored `payment_details/index.html.erb` to reorganize sections for charges, payments, and balances.
- Improved table structure and translation keys to enhance clarity and maintainability.

jjromeo committed:
Merge branch 'RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

jjromeo committed:
[RUBY-2704] Add conditional display for refund link in payment details view

Wrapped the refund link in a conditional check using `@presenter.can_display_refund_link?` to control its visibility based on specific conditions.

jjromeo committed:
Merge branch 'RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

jjromeo committed:
[RUBY-2713] Update `waste_exemptions_engine` gem to latest revision in `Gemfile.lock`

jjromeo committed:
[RUBY-2713] Refactor `ReversePaymentService` to update account balance correctly

- Moved `reversal.save!` inside the transaction block to ensure atomicity.
- Removed the call to `account.update_balance`, directly verifying the updated balance in the test.
- Updated the test to check the account balance change explicitly after reversal.

jjromeo committed:
[RUBY-2713] Refactor payment views to use a shared partial

- Extracted common payment table code into a new partial `_payments_table.html.erb`.
- Updated `record_refunds/index.html.erb` and `record_reversals/index.html.erb` to use the new shared partial, reducing duplication and improving maintainability.

jjromeo committed:
[RUBY-2704] Update `RecordRefundForm` validation for `amount`

Replaced the `amount` validation with a custom `defra_ruby/validators/price` validator to ensure more accurate validation logic.

jjromeo committed:
[RUBY-2704] Add validation for refund amount in `RecordRefundForm`

- Ensure refund `amount` is present and greater than 0.
- Update tests to check for negative amounts and high precision amounts, ensuring they return appropriate error messages.

jjromeo committed:
[RUBY-2704] Update `waste_exemptions_engine` gem to use the `main` branch

Switched the branch for the `waste_exemptions_engine` gem in the `Gemfile` and `Gemfile.lock` from `RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page` to `main`. Updated the revision to reflect the latest commit on the `main` branch.

jjromeo committed:
[RUBY-2713] Add translation prefix to payment tables in refunds and reversals views

- Updated `record_refunds/index.html.erb` and `record_reversals/index.html.erb` to include `translation_prefix` when rendering the payments table.
- Modified `_payments_table.html.erb` to use `translation_prefix` for translation keys, improving maintainability and localization.
- Removed unnecessary comment from `record_reversals.en.yml`.

jjromeo committed:
[RUBY-2713] Update translation keys for refund and reversal links

Changed translation keys in `record_refunds/index.html.erb` and `record_reversals/index.html.erb` to use fully qualified keys instead of relative keys for `refund_payment` and `reverse_payment`.

jjromeo committed:
Merge branch 'RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page' into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

jjromeo committed:
Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

jjromeo committed:
[RUBY-2713] Update schema and Gemfile.lock for payment reversal feature

- Updated `Gemfile.lock` to point to the latest commit in the `waste-exemptions-engine`.
- Modified `db/schema.rb` to reflect changes in the database schema:
  - Removed `reversed_by`, `reversed_at`, and `reversal_id` from the `payments` table.
  - Added `created_by` and `associated_payment_id` to the `payments` table.
  - Added a `

jjromeo committed:
[RUBY-2713] Refactor `ReversePaymentService` for improved reversal handling

- Removed transaction block and updated reversal logic in `ReversePaymentService` to streamline the reversal process.
- Updated schema to replace `reversed_by`, `reversed_at`, and `reversal_id` with `created_by` and `associated_payment_id`.
- Enhanced test coverage in `reverse_payment_service_spec.rb` to validate new reversal handling and ensure the correct attributes are set on reversal creation.
- Updated Gemfile.lock with the latest

jjromeo committed:
Merge branch 'RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

jjromeo committed:
[RUBY-2713] Update Gemfile.lock and refine payment methods in `PaymentDetailsPresenter`

- Updated the `waste-exemptions-engine` gem to a new revision in `Gemfile.lock`.
- Modified `payments` method to exclude both refunds and reversals.
- Updated `refunds_and_reversals` method to include reversals in addition to refunds.

jjromeo committed:
[RUBY-2713] Update waste exemptions engine branch and phonelib gem version

- Changed the `waste_exemptions_engine` gem to use the `main` branch instead of the feature branch.
- Updated `phonelib` gem from version 0.9.3 to 0.10.1.
- Minor formatting adjustments in `reverse_payment_service_spec.rb`.

jjromeo committed:
Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

jjromeo committed:
[RUBY-2713] Remove unnecessary safe navigation operators from controllers and presenters

Refactored `RecordReversalsController` and `PaymentDetailsPresenter` to eliminate redundant safe navigation operators (`&.`), ensuring cleaner and more efficient code.

jjromeo committed:
[RUBY-2713] Update payment ordering and tests for refunds and reversals

- Changed ordering in `PaymentDetailsPresenter` from `date_time` to `created_at`.
- Added `:refund` and `:reversal` traits to the payment factory.
- Updated `payment_details_presenter_spec` to include reversals in the `refunds_and_reversals` method and adjusted tests accordingly.

jjromeo committed:
[RUBY-2713] Update method for displaying payment amount in reversal page

Changed the method from `display_pence_as_pounds_and_cents` to `display_pence_as_pounds_and_pence` for correctly formatting the payment amount on the reverse payment page.

PaulDoyle-EA commented on app/controllers/record_reversals_controller.rb:
```diff
@@ -0,0 +1,48 @@
+    setup_form
+    @payment = @resource.account.payments.find_by(id: params[:payment_id])
+    return if @payment
+
```
Should there be an error if the payment is not found?

PaulDoyle-EA commented on app/presenters/payment_details_presenter.rb:
```diff
@@ -13,28 +13,44 @@ def initialize(registration)
+    registration
+      .account
+      .payments
+      .refundable.any?
```
Are refundable and reversable the same? I would have thought that a partially-reversed payment would be refundable but not reversable.

PaulDoyle-EA commented on app/services/reverse_payment_service.rb:
```diff
@@ -0,0 +1,36 @@
+
+    true
+  rescue StandardError => e
+    Rails.logger.error "#{e.class} error processing reversal for payment #{payment.id}"
```
Also log the error message

PaulDoyle-EA commented on app/views/payment_details/index.html.erb:
```diff
@@ -165,8 +165,15 @@
+      </p>
+    <% end %>
+
+    <% if @presenter.can_display_reversal_link? %>
```
This method should be on ActionLinksHelper, not on the presenter

PaulDoyle-EA commented on app/views/record_refunds/index.html.erb:
```diff
@@ -1,62 +1,11 @@
+  <%= link_to new_registration_record_refund_path(reference: @resource.reference, payment_id: payment.id) do %>
+    <%= t("record_refunds.index.refund_payment") %>
+    <span class="govuk-visually-hidden">
+      <%= "ALT" %>
```
What is this for?

jjromeo replied:
Placeholder for screenreader content, but I guess may not be needed

PaulDoyle-EA commented on app/views/record_reversals/index.html.erb:
```diff
@@ -0,0 +1,11 @@
+  <%= link_to new_registration_record_reversal_path(reference: @resource.reference, payment_id: payment.id) do %>
+    <%= t("record_reversals.index.reverse_payment") %>
+    <span class="govuk-visually-hidden">
+      <%= "ALT" %>
```
And this?

jjromeo replied:
Removed

PaulDoyle-EA commented on app/views/record_reversals/new.html.erb:
```diff
@@ -0,0 +1,47 @@
+      <div class="govuk-form-group">
+        <h2 class="govuk-heading-m"><%= t(".amount_to_reverse") %></h2>
+        <p class="govuk-body">
+          £<%= display_pence_as_pounds_and_pence(pence: @payment.payment_amount) %>
```
The "£" should be added by the helper - see `display_pence_as_pounds_sterling_and_pence`.

jjromeo replied:
Done

PaulDoyle-EA commented on app/views/shared/_payments_table.html.erb:
```diff
@@ -0,0 +1,55 @@
+        <% payments.each do |payment| %>
+          <tr class="govuk-table__row">
+            <td class="govuk-table__cell">
+              <%= payment.created_at.to_date.to_formatted_s(:day_month_year_slashes) %>
```
Can this go in a presenter? Or is that follow-on work?

jjromeo replied:
Follow-on work

PaulDoyle-EA commented on app/views/shared/_payments_table.html.erb:
```diff
@@ -0,0 +1,55 @@
+              <%= t("#{translation_prefix}.payment_types.#{payment.payment_type}") %>
+            </td>
+            <td class="govuk-table__cell govuk-table__cell--numeric">
+              £<%= display_pence_as_pounds_and_pence(pence: payment.payment_amount, hide_pence_if_zero: false) %>
```
As above - see sterling helper

PaulDoyle-EA commented on db/schema.rb:
```diff
@@ -138,6 +138,15 @@
     t.string "sent_to"
   end
 
+  create_table "companies", force: :cascade do |t|
```
What is this for?

PaulDoyle-EA commented on db/schema.rb:
```diff
@@ -16,7 +16,7 @@
   enable_extension "tsm_system_rows"
 
   create_table "accounts", force: :cascade do |t|
-    t.bigint "registration_id", null: false
```
I've seen this locally too, I don't think `null: false` should be dropped.

jjromeo committed:
[RUBY-2713] Make `registration_id` non-nullable in `accounts` and remove `companies` table

- Updated `accounts` table schema to require `registration_id`.
- Dropped the `companies` table from the schema.

jjromeo committed:
[RUBY-2713] Add reversal link logic and update payment display

- Introduced `can_display_reversal_link?` method in `ActionLinksHelper` to determine if a reversal link should be shown.
- Enhanced error logging in `ReversePaymentService` to include error messages.
- Updated `payment_details/index.html.erb` to use the new reversal link helper method.
- Removed unnecessary `span` elements from `record_refunds/index.html.erb` and `record_reversals/index.html.erb`.
-

jjromeo committed:
Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

jjromeo committed:
[RUBY-2713] Fix error message translation key and update user role in specs

- Corrected the translation key for the error message in `RecordReversalForm` to use the correct path.
- Updated the user role from `:super_agent` to `:admin_agent` in `record_reversals_spec.rb` for consistency.

jjromeo committed:
[RUBY-2713] Refactor recording reversals to go with payment presenter
refactor
Squashed commit of the following:

commit ee63ed0d7778bb151f211e6b3c112b8c5a29e2d6
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Thu Dec 5 17:11:56 2024 +0000

    [RUBY-2713] Refactor `ActionLinksHelper` specs for `can_display_reversal_link?`

    - Changed `registration` from `build` to `create` to ensure persistence in tests.
    - Added setup for account payments using `before` block to improve test clarity.
    - Introduced a new context for handling non-reversible payments to enhance test coverage.

commit dcde7a24cda75b820fbcd56af565ffc95f9d164d
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Thu Dec 5 16:46:13 2024 +0000

    [RUBY-2713] Refactor `ActionLinksHelper` and update database schema

    - Updated `ActionLinksHelper` methods to use `reverseable` instead of `refundable`.
    - Modified `schema.rb` to reflect changes in the database, including making `registration_id` nullable in `accounts`, adding `companies` and `waste_activities` tables, and introducing new fields and indexes.
    - Enhanced `ActionLinksHelper` specs with tests for `can_display_reversal_link?`.
    - Refactored `

commit 4b3acbba3adc1c46466b1d106c9a89e6e9f2bb3b
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Thu Dec 5 16:19:36 2024 +0000

    [RUBY-2713] Refactor `RecordReversalForm` and add tests

    - Updated `RecordReversalForm` to use `ReversePaymentService.run` directly, removing unnecessary logging.
    - Improved error handling in `record_refund_service.rb` and `reverse_payment_service.rb` by using safe navigation (`&.`) for payment ID logging and notifications.
    - Added `record_reversal_form_spec.rb` to test the `RecordReversalForm`, covering scenarios with valid and invalid parameters, ensuring correct service calls and

commit 9c0666762ee0672398c889cd57ae2ab9deb117f8
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Thu Dec 5 15:58:57 2024 +0000

    [RUBY-2713] Refactor: Remove `PaymentDetailsPresenter` and update refund handling

    - Deleted `payment_details_presenter.rb` to streamline code and remove unused presenter logic.
    - Modified `RecordRefundService` to include `associated_payment` for better refund tracking.
    - Enhanced test coverage in `payment_spec.rb` by adding tests for the `.reverseable` method.
    - Updated `record_refund_service_spec.rb` to test the inclusion of `associated_payment` in refund records.

commit 9e0c76eb16e72277d136e7d25e1f553820b4d40a
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Thu Dec 5 15:04:18 2024 +0000

    [RUBY-2713] Refactor RecordReversalsController to use PaymentPresenter

    - Updated the `index` action to map `@payments` using `PaymentPresenter`.
    - Modified the `new` action to initialize `@presenter` with `PaymentPresenter` if `@payment` is present.
    - Ensured `@presenter` is set before rendering the `new` action in case of errors.

commit 09e9703de174b20cc0b2dc97433425ee07227106
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Thu Dec 5 14:32:39 2024 +0000

    [RUBY-2713] Refactor `Payment` spec to improve readability and structure

    - Nested `RSpec.describe` under `WasteExemptionsEngine` module for clarity.
    - Improved indentation and formatting for better readability.
    - Added explicit context descriptions for `#maximum_refund_amount` tests.
    - Ensured consistent use of `let` and `before` blocks for setup.

commit 2ba3f5df3c3fe4d76219079cd3130c802e5dadc1
Merge: 187b8f6 bf3902b
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Thu Dec 5 14:29:37 2024 +0000

    Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2713-with-presenters-refactor

commit bf3902b0e24c9a74d9b4788dd5688ef65d97e9a4
Author: Jerome Pratt <jerome.pratt@defra.gov.uk>
Date:   Wed Dec 4 16:41:10 2024 +0000

    [RUBY-2704] Fix presenters bug in `RecordRefundsController` and views (#1543)

    * [RUBY-2704] Fix presenters bug in `RecordRefundsController` and views

    - Updated `RecordRefundsController` to use `PaymentPresenter` for payments, ensuring consistent data presentation.
    - Modified `PaymentPresenter` to include `ApplicationHelper` and added methods for formatted display of payment attributes.
    - Adjusted views to utilize presenter methods for displaying payment details, improving code readability and maintainability.

    * [RUBY-2704] Fix nullable chain issue in `RecordRefundsController`

    - Removed unnecessary safe navigation operators when accessing `@resource.account.payments`.
    - Ensured `PaymentPresenter` is initialized in the `create` action when rendering the `new` view in case of errors.

commit 7a2ee37944e4293b0e6c695d1e66f6bb95e90908
Author: Jerome Pratt <jerome.pratt@defra.gov.uk>
Date:   Wed Dec 4 13:57:29 2024 +0000

    Fixes to Rounding defect and misleading refund amount (#1542)

commit cd455c569ff8c4292603a53e38573f35d46a86cf
Author: Jerome Pratt <jerome.pratt@defra.gov.uk>
Date:   Wed Dec 4 11:56:31 2024 +0000

    RUBY 2704 payment details presenter refactor (#1534)

    * [RUBY-2704] Refactor payment details presenter and update related components

    - Changed the branch for the `waste_exemptions_engine` gem in the `Gemfile` and `Gemfile.lock` to `RUBY-2704-payment-details-presenter-refactor` and updated the revision.
    - Upgraded `phonelib` gem from version 0.9.3 to 0.10.1.
    - Refactored `PaymentDetailsController#index` to use `AccountPresenter` instead of `PaymentDetailsPresenter`.
    - Added a new helper module `ApplicationHelper` with a `format_date` method for date formatting.
    - Updated `FinanceDetailsHelper` to include `ActionView::Helpers::NumberHelper`.
    - Introduced `AccountPresenter`, `OrderPresenter`, and `PaymentPresenter` classes:
      - `AccountPresenter` handles balance, successful payments, refunds and reversals, and orders.
      - `OrderPresenter` provides a method to retrieve exemption codes.
      - `PaymentPresenter` formats payment type and amount.
    - Removed `PaymentDetailsPresenter` and its corresponding spec file `payment_details_presenter_spec.rb`.
    - Updated `payment_details/index.html.erb` to use new presenter methods for displaying payment and refund details.
    - Removed unnecessary payment type translations from `payment_details.en.yml`.
    - Added a new empty helper module `ApplicationHelpers`.
    - These changes improve the separation of concerns and enhance the maintainability of the codebase by introducing dedicated presenter classes to handle data presentation logic.

    * [RUBY-2704] Refactor payment details presenter

    - Removed `application_helpers.rb` as it is no longer needed.
    - Corrected spacing in `account_presenter.rb` for consistency in block syntax.
    - Added missing space in `payment_presenter.rb` after the frozen string literal comment.

    * [RUBY-2703] Add `charge_adjustments` table to schema

    - Updated the schema version to 2024_11_18_143709.
    - Created `charge_adjustments` table with columns: `account_id`, `amount`, `adjustment_type`, `reason`, `created_at`, and `updated_at`.
    - Added an index on `account_id` for the `charge_adjustments` table.
    - Added a foreign key constraint linking `charge_adjustments` to the `accounts` table.

    * [RUBY-2704] Refactor `AccountPresenter` for improved code reuse

    - Simplified `successful_payments` and `refunds_and_reversals` methods by using `super` to leverage inherited behavior.
    - Renamed `orders` to `sorted_orders` to clarify method purpose and ensure consistent method naming.

    * [RUBY-2704] Remove `payment_details_presenter_spec.rb` file

    The `payment_details_presenter_spec.rb` file has been deleted as part of the refactoring process. This change may be due to the removal of obsolete tests or the restructuring of the test suite to better align with the updated application code.

    * [RUBY-2704] Refactor `PaymentDetailsPresenter`: Use `sorted_orders` in view

    Updated the payment details index view to use `@presenter.sorted_orders` instead of `@presenter.orders` to ensure orders are displayed in sorted order.

    * [RUBY-2704] Update `Gemfile.lock` to latest commit of `waste-exemptions-engine` for branch `RUBY-2704-payment-details-presenter-refactor`.

    * [RUBY-2704] Refactor `sorted_orders` method in `AccountPresenter`

    Updated the `sorted_orders` method to order `orders` by `created_at` in descending order before mapping them to `OrderPresenter` instances.

    * [RUBY-2704] Add payment model from waste exemptions engine with scopes

    This moves the scopes from the engine that are only used in the back
    office into the back office

    * [RUBY-2704] Add `Account` model with delegation to `payments` methods in WasteExemptionsEngine

    These delegations are only used in back office

    * [RUBY-2704] Update Gemfile to use main branch for waste_exemptions_engine

    - Changed the branch from `RUBY-2704-payment-details-presenter-refactor` to `main` in the Gemfile and Gemfile.lock for the waste_exemptions_engine.
    - Updated the revision in Gemfile.lock to match the main branch.
    - Removed an unnecessary newline in `account.rb`.

    * [RUBY-2704] Resolve merge conflict in `account.rb` by updating the require statement.

    * [RUBY-2704] Refactor `Payment` spec to include registration in account setup

    Updated the `let` block in the `#successful_payments` spec to create an account with an associated registration, ensuring the test setup accurately reflects the required data structure.

    * [RUBY-2704] Refactor `Payment` spec to DRY up `account` setup

    Moved the `let(:account)` declaration to a higher scope to remove duplication in the `#successful_payments` and `#refunds_and_reversals` tests.

    * [RUBY-2704] Refactor `payment_spec.rb` to improve code readability

    Added a blank line for better separation and readability in the `RSpec.describe` block for `WasteExemptionsEngine::Payment`.

commit 187b8f6818af286df4d427c06b6e2c275346a499
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Wed Dec 4 09:52:39 2024 +0000

    [RUBY-2713] Remove `payment_details_presenter_spec.rb` as part of presenters refactor

    The spec file for `PaymentDetailsPresenter` has been deleted, likely due to changes in the presenter structure or testing strategy.

commit 98efc0bbed85e266444818278e770bd3c94b4e1e
Merge: 9b6803d 8ca1360
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Dec 3 20:17:24 2024 +0000

    Merge branch 'RUBY-2704-payment-details-presenter-refactor' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2713-with-presenters-refactor

commit 8ca1360142df835ed890e417d085222f6471b884
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Dec 3 17:04:05 2024 +0000

    [RUBY-2704] Refactor `payment_spec.rb` to improve code readability

    Added a blank line for better separation and readability in the `RSpec.describe` block for `WasteExemptionsEngine::Payment`.

commit 8983de0bace62daa4e2beefb1f2aaf2694ae932b
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Dec 3 16:58:35 2024 +0000

    [RUBY-2704] Refactor `Payment` spec to DRY up `account` setup

    Moved the `let(:account)` declaration to a higher scope to remove duplication in the `#successful_payments` and `#refunds_and_reversals` tests.

commit c0a492bd79e97f0d26f6a1644b3fa8eb45139d99
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Dec 3 16:38:45 2024 +0000

    [RUBY-2704] Refactor `Payment` spec to include registration in account setup

    Updated the `let` block in the `#successful_payments` spec to create an account with an associated registration, ensuring the test setup accurately reflects the required data structure.

commit ecd818673ffb536310fd101ace44b4477c584ac4
Merge: 253de50 46fd48a
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Dec 3 16:22:08 2024 +0000

    Merge branch 'RUBY-2704-payment-details-presenter-refactor' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-payment-details-presenter-refactor

commit 253de502faf2e2cc93ab356b7a0232392064a3ae
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Dec 3 16:21:53 2024 +0000

    [RUBY-2704] Resolve merge conflict in `account.rb` by updating the require statement.

commit 46fd48a6659c3f5de523794e5a95940f84712843
Merge: d503a51 6e913af
Author: Jerome Pratt <jerome.pratt@defra.gov.uk>
Date:   Tue Dec 3 16:08:14 2024 +0000

    Merge branch 'main' into RUBY-2704-payment-details-presenter-refactor

commit d503a51c9ed861919032b2109a12df9efbc50203
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Dec 3 16:04:04 2024 +0000

    [RUBY-2704] Update Gemfile to use main branch for waste_exemptions_engine

    - Changed the branch from `RUBY-2704-payment-details-presenter-refactor` to `main` in the Gemfile and Gemfile.lock for the waste_exemptions_engine.
    - Updated the revision in Gemfile.lock to match the main branch.
    - Removed an unnecessary newline in `account.rb`.

commit 81fa84668257b219d6bcc9b852b12dbf29448b5c
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Dec 3 15:54:30 2024 +0000

    [RUBY-2704] Add `Account` model with delegation to `payments` methods in WasteExemptionsEngine

    These delegations are only used in back office

commit 48f6f2d292fdb5441eb7a0fc38081f1e448f3fd4
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Dec 3 15:35:43 2024 +0000

    [RUBY-2704] Add payment model from waste exemptions engine with scopes

    This moves the scopes from the engine that are only used in the back
    office into the back office

commit 9b6803da201ccd9d2221d01a339a6e6081f38821
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Dec 3 15:02:35 2024 +0000

    [RUBY-2713] Refactor presenters to include `ApplicationHelper` and update date formatting

    - Added `ApplicationHelper` inclusion in `BasePresenter` to utilize shared methods.
    - Updated `PaymentPresenter` to format `created_at` using `format_date`.
    - Modified payments table view to use the new `created_at` method for date formatting.

commit 7b82f0bbcfc6908fb5688f9a4ad16531b6acf04f
Merge: eaf21db fd8a3c9
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Dec 3 10:00:56 2024 +0000

    Merge branch 'RUBY-2704-payment-details-presenter-refactor' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2713-with-presenters-refactor

commit fd8a3c921a0d70ca295b9f74412d090af120515e
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Mon Dec 2 15:36:36 2024 +0000

    [RUBY-2704] Refactor `sorted_orders` method in `AccountPresenter`

    Updated the `sorted_orders` method to order `orders` by `created_at` in descending order before mapping them to `OrderPresenter` instances.

commit a450ed880f92d1b51cfd0f6f66c54776a62b48d3
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Fri Nov 29 14:14:23 2024 +0000

    [RUBY-2704] Update `Gemfile.lock` to latest commit of `waste-exemptions-engine` for branch `RUBY-2704-payment-details-presenter-refactor`.

commit b0d76697561e72457946eb8fb55ff644737f3390
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Fri Nov 29 14:00:59 2024 +0000

    [RUBY-2704] Refactor `PaymentDetailsPresenter`: Use `sorted_orders` in view

    Updated the payment details index view to use `@presenter.sorted_orders` instead of `@presenter.orders` to ensure orders are displayed in sorted order.

commit 4f0a48e384bec0a13ffd36bc831b5c91a30134a4
Merge: 8e99833 7cdd5e7
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Fri Nov 29 13:50:22 2024 +0000

    Merge branch 'RUBY-2704-payment-details-presenter-refactor' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-payment-details-presenter-refactor

commit 8e99833ca0c9260c256107c488a8283e7e0881eb
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Fri Nov 29 13:48:14 2024 +0000

    [RUBY-2704] Remove `payment_details_presenter_spec.rb` file

    The `payment_details_presenter_spec.rb` file has been deleted as part of the refactoring process. This change may be due to the removal of obsolete tests or the restructuring of the test suite to better align with the updated application code.

commit 7cdd5e7808fd4178fd168c80e8416d62f581bcf3
Merge: 02b53ff e877db1
Author: Jerome Pratt <jerome.pratt@defra.gov.uk>
Date:   Fri Nov 29 13:47:48 2024 +0000

    Merge branch 'main' into RUBY-2704-payment-details-presenter-refactor

commit 02b53ff90aeb0f766d36bcefa2d4b059140dae2c
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Fri Nov 29 12:45:57 2024 +0000

    [RUBY-2704] Refactor `AccountPresenter` for improved code reuse

    - Simplified `successful_payments` and `refunds_and_reversals` methods by using `super` to leverage inherited behavior.
    - Renamed `orders` to `sorted_orders` to clarify method purpose and ensure consistent method naming.

commit 7cba0530247f8e8bc26b544042c0854855d20e0b
Merge: ba2ca08 4a87ab6
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Fri Nov 29 10:30:40 2024 +0000

    Merge branch 'RUBY-2703-update-schema' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-payment-details-presenter-refactor

commit ba2ca0824b08d665dc6468414aafc6f9b1551350
Merge: 263cef2 a677d87
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Fri Nov 29 10:30:26 2024 +0000

    Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-payment-details-presenter-refactor

commit 4a87ab6a6f8ea22dcb3f6b37c6cecb61c03c52f0
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Fri Nov 29 10:16:59 2024 +0000

    [RUBY-2703] Add `charge_adjustments` table to schema

    - Updated the schema version to 2024_11_18_143709.
    - Created `charge_adjustments` table with columns: `account_id`, `amount`, `adjustment_type`, `reason`, `created_at`, and `updated_at`.
    - Added an index on `account_id` for the `charge_adjustments` table.
    - Added a foreign key constraint linking `charge_adjustments` to the `accounts` table.

commit 263cef2f9cc21cdea68fa75db055fd350b84761a
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Thu Nov 28 17:04:34 2024 +0000

    [RUBY-2704] Refactor payment details presenter

    - Removed `application_helpers.rb` as it is no longer needed.
    - Corrected spacing in `account_presenter.rb` for consistency in block syntax.
    - Added missing space in `payment_presenter.rb` after the frozen string literal comment.

commit c2cae63251e7051683152f919a7581c40cfc71aa
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Thu Nov 28 17:03:42 2024 +0000

    [RUBY-2704] Refactor payment details presenter and update related components

    - Changed the branch for the `waste_exemptions_engine` gem in the `Gemfile` and `Gemfile.lock` to `RUBY-2704-payment-details-presenter-refactor` and updated the revision.
    - Upgraded `phonelib` gem from version 0.9.3 to 0.10.1.
    - Refactored `PaymentDetailsController#index` to use `AccountPresenter` instead of `PaymentDetailsPresenter`.
    - Added a new helper module `ApplicationHelper` with a `format_date` method for date formatting.
    - Updated `FinanceDetailsHelper` to include `ActionView::Helpers::NumberHelper`.
    - Introduced `AccountPresenter`, `OrderPresenter`, and `PaymentPresenter` classes:
      - `AccountPresenter` handles balance, successful payments, refunds and reversals, and orders.
      - `OrderPresenter` provides a method to retrieve exemption codes.
      - `PaymentPresenter` formats payment type and amount.
    - Removed `PaymentDetailsPresenter` and its corresponding spec file `payment_details_presenter_spec.rb`.
    - Updated `payment_details/index.html.erb` to use new presenter methods for displaying payment and refund details.
    - Removed unnecessary payment type translations from `payment_details.en.yml`.
    - Added a new empty helper module `ApplicationHelpers`.
    - These changes improve the separation of concerns and enhance the maintainability of the codebase by introducing dedicated presenter classes to handle data presentation logic.

commit f31187519fb1b485f78a85c0b4e0dfcdd35c107e
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Wed Nov 27 11:36:28 2024 +0000

    [RUBY-2704] Fix indentation in `payment_details_presenter_spec.rb` for balance expectation test.

commit ed2355d3f02d4a629eccde62b95ea6938e91fad3
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Wed Nov 27 10:01:57 2024 +0000

    [RUBY-2704] Update balance calculation in `PaymentDetailsPresenter` spec

    - Refactored balance expectation to use `WasteExemptionsEngine::CurrencyConversionService.convert_pence_to_pounds` for converting account balance from pence to pounds.

commit b8ec28fcfa9db14cef23928196a75048c80b6878
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Nov 26 13:12:59 2024 +0000

    [RUBY-2704] Refactor `RecordRefundForm` to directly use `RecordRefundService`

    Removed the dependency injection of `RecordRefundService` in the `submit` method of `RecordRefundForm` to simplify the code by directly calling `RecordRefundService.run`.

commit b53d8467c9dd7403b882f1400c87cf83a69ecda5
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Nov 26 11:15:53 2024 +0000

    [RUBY-2704] Refactor `RecordRefundForm` to improve amount handling

    - Removed direct conversion of amount to pence in `submit` method.
    - Introduced `amount_in_pence` method for consistent conversion and validation.
    - Simplified validation logic for refund amount limits.

commit 0807681e6a9cec6076c7b2bc12a8c8a8cee08990
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Nov 26 11:03:07 2024 +0000

    [RUBY-2704] Remove division by 100 for balance and payment amount checks

    Adjusted the `@balance` and `amount_within_limits` methods in `RecordRefundForm` to use the raw balance and payment amount values without dividing by 100. This ensures that the calculations are accurate and consistent with the stored values.

commit a7fabb25f00efe2549940f6f931ec4bce0b42cf8
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Nov 26 10:18:24 2024 +0000

    [RUBY-2704] Convert refund amounts from pounds to pence

    - Updated `RecordRefundForm` to convert refund amounts from pounds to pence using `CurrencyConversionService`.
    - Modified `RecordRefundService` to handle amounts in pence instead of pounds.
    - Adjusted related tests in `record_refund_service_spec.rb` to use amounts in pence.

commit 5093bf4504953d4e4b286ba1e36745818c2f0dc5
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Fri Nov 22 13:18:33 2024 +0000

    [RUBY-2704] Update `waste_exemptions_engine` gem to latest revision in `Gemfile.lock`

commit cb5bed5d5006993ab76e12834a1e8571a43985a0
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Thu Nov 21 15:57:33 2024 +0000

    [RUBY-2704] Update refund page to use dynamic payment type and refactor form fields

    - Modified the refund form to display dynamic payment types based on `@payment.payment_type`.
    - Refactored form fields to use `govuk_text_field` and `govuk_text_area` to fix issue with errors not linking back to ifelds
    - Updated locale file to include additional payment types and improve translation keys.

commit c4032c261a7a416e3d7b5c7b63814670368710bd
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Wed Nov 20 17:18:12 2024 +0000

    [RUBY-2704] Refactor `RecordRefundForm` to improve payment validation

    - Removed `payment_exists` validation method and replaced it with `check_payment_exists?` method, which returns a boolean.
    - Ensured `check_payment_exists?` is called before accessing `payment` to prevent errors.
    - Simplified balance calculation by removing unnecessary checks and directly dividing `payment.account.balance` by 100.

commit c2ce7ed1092b6d1b4da04b3d0f77d00155bfb2d8
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Wed Nov 20 14:12:24 2024 +0000

    [RUBY-2704] Refactor `RecordRefundForm` to improve nil handling

    - Extracted `account` from `payment` to safely handle potential nil values before accessing `balance`.

commit d8c9b6b43b849e94a551dc46a1275c05142f736b
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Wed Nov 20 13:31:30 2024 +0000

    [RUBY-2704] Rubocop fix

    - reduce safe navigation chain

commit b377ee28876232b4690e9ef7ca01a3e04603a46f
Merge: 3327fd4 5d8cbc8
Author: Jerome Pratt <jerome.pratt@defra.gov.uk>
Date:   Wed Nov 20 13:23:20 2024 +0000

    Merge branch 'main' into RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page

commit 3327fd43d887362fd2d869d7fe4b7138fdc3d5ec
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Wed Nov 20 09:54:01 2024 +0000

    [RUBY-2704] set date in payments table to date_time to fix issue in table

commit f2b783668b584f23fcbf6598624ea7c0d497843e
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Nov 19 15:21:22 2024 +0000

    [RUBY-2704] Refactor `RecordRefundForm` validation logic

    - Simplified validation messages by removing inline I18n translations and using error symbols.
    - Removed `reason_present_in_comments` method and replaced it with a direct presence validation on `comments`.
    - Updated the locale file to reflect changes in validation message structure, moving messages to `activemodel.errors` format for better consistency and maintainability.

commit f15a84822ad85fc5edde73d6a95b91da2591d4ef
Author: Jerome Pratt <jerome.pratt@hey.com>
Date:   Tue Nov 19 11:03:19 2024 +0000

    [RUBY-2704] Fix bug where refund can be higher than balance

    - Enhanced validation to check if refund amount is greater than or equal to the balance.
    - Removed redundant balance assignment in test setup.
    - Added a test case to ensure refund amount does not exceed balance.

jjromeo committed:
[RUBY-2713] Update SonarCloud GitHub Action in CI workflow

Changed the GitHub Action for SonarCloud analysis from `sonarcloud-github-action` to `sonarqube-scan-action` in the CI workflow configuration file.

jjromeo committed:
Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

jjromeo committed:
Merge branch 'main' into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

jjromeo committed:
Merge branch 'main' into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

jjromeo committed:
Merge branch 'main' into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

jjromeo committed:
Merge branch 'main' into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

jjromeo committed:
Merge branch 'main' into RUBY-2713-wex-charging-bo-3-1-5-reverse-a-payment-page

PaulDoyle-EA commented on db/schema.rb:
```diff
@@ -10,13 +10,13 @@
 
   create_table "accounts", force: :cascade do |t|
-    t.bigint "registration_id", null: false
+    t.bigint "registration_id"
```
Again, should `null: false` be retained? I keep seeing this locally too.

PaulDoyle-EA commented on db/schema.rb:
```diff
@@ -148,6 +148,15 @@
   end
 
+  create_table "companies", force: :cascade do |t|
+    t.string "name"
```
Unclear to me why this change is part of this PR

PaulDoyle-EA commented on db/schema.rb:
```diff
@@ -157,7 +166,9 @@
     t.text "guidance"
     t.boolean "hidden", default: false
     t.bigint "band_id"
+    t.bigint "waste_activity_id"
```
Likewise

PaulDoyle-EA commented on db/schema.rb:
```diff
@@ -399,6 +413,9 @@
     t.boolean "temp_check_your_answers_flow"
     t.string "temp_company_no"
     t.string "temp_payment_method"
+    t.text "temp_waste_activities", default: [], array: true
```
Likewise

PaulDoyle-EA commented on db/schema.rb:
```diff
@@ -455,6 +472,14 @@
     t.index ["item_type", "item_id"], name: "index_versions_on_item_type_and_item_id"
   end
 
+  create_table "waste_activities", force: :cascade do |t|
```
And this

jjromeo committed:
[RUBY-2713] Add permissions check for reverse payment link

- Updated the view to include a permission check (`can?(:reverse_payment, @resource)`) before displaying the reverse payment link.
- Enhanced the test suite to cover scenarios where a user does or does not have the `reverse_payment` permission, ensuring the link is displayed or hidden accordingly.

jjromeo committed:
[RUBY-2713] Refactor `payment_details_spec.rb` for clarity and organization

- Removed the unused `payment` let variable.
- Introduced `i18n_actions_section` let variable earlier for better readability.
- Reorganized the "reverse payment link" tests into a separate describe block for clarity.
- Adjusted expectations to directly access `account.payments.first.reference` for improved test accuracy.

jjromeo committed:
[RUBY-2713] Refactor payment details spec for clarity and efficiency

- Removed the unused `added_payments` variable.
- Introduced `i18n_actions_section` variable outside user context for reuse.
- Simplified setup in the "reverse payment link" test by directly using a `payment` variable instead of a list.
- Ensured `sign_in` is called before account payments are modified and the page is accessed.
